components/raftstore/src/store/util.rs

```diff
@@ -863,8 +853,8 @@ impl<'a> ChangePeerI for &'a ChangePeerV2Request {
             })
             .collect();
 
-        if self.get_transition() != eraftpb::ConfChangeTransition::Auto {
-            cc.set_transition(self.get_transition());
+        if self.get_is_unsafe_recovery_remove_peers() {
+            cc.set_transition(eraftpb::ConfChangeTransition::Auto);
```

Comment:
Why set to `ConfChangeTransition::Auto`? it will turn on the `auto_leave`

Reply:
The idea is to not use the joint consensus for updating the peer list. By explicitly set it to `Auto`, Raft considers this v2 conf change request as `simple` thus does not enter the joint state. In other words, unsafe recovery only utilize conf change v2 for removing multiple peers simultaneously.  How do you think of this idea?

Reply:
It will enter joint state as long as there are multiple `confchange`, see: 

https://github.com/tikv/raft-rs/blob/2357cb22760719bcd107a90d1e64ef505bdb1e15/proto/src/confchange.rs#L134-L141

https://github.com/tikv/raft-rs/blob/2357cb22760719bcd107a90d1e64ef505bdb1e15/src/raft.rs#L2738-L2744

Reply:
Indeed, I misread the code. The test passes because the raftstore group conf, after the changes made in this PR, is changed once the `enter_joint` message is applied, while the raft internal conf is left in joint state forever.

Seems that we are in a dilemma here:
* Removing a voter is prohibited (see `raftstore/src/store/peer.rs:check_conf_change` and `raftstore/src/store/fsm/apply.rs:apply_conf_change`)
* If we bypasses the `voter removing check` like this PR already did, the `leave joint` still won't pass, since it does not support any peer role other than "incoming voter" and "demoting voter" (see `raftstore/src/store/fsm/apply.rs:apply_leave_joint`)
* Demote a voter first, then remove it is also not allowed, since "learner only change" is prohibited (see `raftstore/src/store/peer.rs:check_conf_change()/only_learner_change`)

My suggestion is that, change the `leave joint` logic as well. If it is an unsafe recovery removing peer request, do not change peer role as other `leave joint` request does, but still call the `raft leave joint`. What do you think?

Reply:
> Demote a voter first, then remove it is also not allowed, since "learner only change" is prohibited

Actually, after demoting the voters of the failed store, the region is recovered, and the remaining learners can be clean up by PD one by one later. 

Besides, "multiple learner only changes"  is actually safe on theory but restricted by the current implementation (namely by the condition of whether enter joint) and there is an optimization talked before that not enter joint state for learner-only changes but not yet implemented due to no required use case, we can also implement it if needed.

 > change the leave joint logic as well. If it is an unsafe recovery removing peer request, do not change peer role as other leave joint request does, but still call the raft leave joint

This approch also okay to me (although I more perfer the demote voters approch, which bring less change to the joint consensus mechanism)